File: Matrix/Medium/find_nth_element_in_spiral_matrix.py
# You are required to complete this fucntion
# Function should return the an integer
# https://practice.geeksforgeeks.org/problems/find-nth-element-of-spiral-matrix/1
# Given a matrix with n rows and m columns. Your task is to find the kth element which is obtained
# while traversing the matrix spirally. You need to complete the method findK which takes four
# arguments the first argument is the matrix A and the next two arguments will be n and m denoting
# the size of the matrix A and then the forth argument is an integer  k denoting the kth element . 
# The function will return the kth element obtained while traversing the matrix spirally.
def findK(matrix, r, c, tar):
    k = 0
    while(r > k and c > k):
        for j in range(k, c):
            tar -= 1
            if(tar == 0):
                return matrix[k][j]
            
        for i in range(k+1, r):
            tar -= 1
            if(tar == 0):
                return matrix[i][c-1]
            
        j = c-2
        while(k >= k and r != k+1):
            tar -= 1
            if(tar == 0):
                return matrix[r-1][j]
                
            j -= 1
            
        i = r -2
        while(i > k and c != k + 1):
            tar -= 1
            if(tar == 0):
                return matrix[i][k]
                
            i -= 1 
            
        r -= 1
        c -= 1
        k += 1
        
    return -1


#{ 
 # Driver Code Starts
# Your code goes here
if __name__=='__main__':
    t = int(input())
    for i in range(t):
        n = list(map(int, input().strip().split()))
        arr = list(map(int, input().strip().split()))
        matrix = [[0 for i in range(n[1])]for j in range(n[0])]
        c=0
        for i in range(n[0]):
            for j in range(n[1]):
                matrix[i][j] = arr[c]
                c+=1
        print(findK(matrix, n[0], n[1], n[2]))
# } Driver Code Ends

File: Matrix/Medium/find_nth_element_in_spiral_matrix.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;
#define MAX 1000
int findK(int [MAX][MAX],int ,int ,int );
int main()
{
    int T;
    cin>>T;
  
    while(T--)
    {
        int n,m;
        int k=0;
        //cin>>k;
        cin>>n>>m>>k;
        int a[MAX][MAX];
        
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                cin>>a[i][j];
            }
        }
        cout<<findK(a,n,m,k)<<endl;
        
       
    }
}
// } Driver Code Ends

// https://practice.geeksforgeeks.org/problems/find-nth-element-of-spiral-matrix/1
// Given a matrix with n rows and m columns. Your task is to find the kth element which is obtained
// while traversing the matrix spirally. You need to complete the method findK which takes four
// arguments the first argument is the matrix A and the next two arguments will be n and m denoting
// the size of the matrix A and then the forth argument is an integer  k denoting the kth element . 
// The function will return the kth element obtained while traversing the matrix spirally.
/*You are required to complete this method*/
int findK(int matrix[MAX][MAX], int r, int c, int tar)
{
    int k = 0;
    while(r > k && c > k)
    {
        for(int j = k; j < c; j++)
            if(--tar == 0) return matrix[k][j];
            
        for(int i = k + 1; i < r; i++)
            if(--tar == 0) return matrix[i][c-1];
            
        for(int j = c-2; j >= k && r!= k+1; j--)
            if(--tar == 0) return matrix[r-1][j];
            
        for(int i = r-2; i > k && c != k+1; i--)
            if(--tar == 0) return matrix[i][k];
            
        r--;
        c--;
        k++;
    }
    return -1;
}

